### PR TITLE
fix: Use correct type for let expression generated by datatype update

### PR DESCRIPTION
**Pull request by cpitclaudel**
Closes GH-1479.

Co-authored-by: @RustanLeino

### CODE REVIEW

File: Test/git-issues/git-issue-1479a.dfy [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+type T
+datatype BaseType = BaseType(t: T)
+
+predicate P(d: BaseType)
+type SubsetType = d: BaseType | P(d) witness *
+
+method m0(dp: SubsetType)
+
+method m1(dp: SubsetType, t: T) {
+  var dp': SubsetType := [dp][0].(t := t);
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1479a.dfy:13]: Add a comment here, saying we expect an error on this line.

File: Test/git-issues/git-issue-1479b.dfy [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  predicate i() { v == 0 }
+}
+
+datatype S = S(rep: map<nat, R>) {
+  function launder(r: R) : R { r }
+
+  function cpy(r: nat, v: nat): R
+    requires r in rep
+  {
+    launder(rep[r].(v := v))
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1479b.dfy:16]: Add a comment saying this is an error.

File: Test/git-issues/git-issue-1479a.dfy [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+type T
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-1479a.dfy:4]: It's fine to have the two test files, `git-issue-1479a.dfy` and `git-issue-1479b.dfy`, but if (for this PR or some future PRs) you want to combine them into one file and there are name clashes, then you can put each example in a separate `module` in the one test file.

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Test/git-issues/git-issue-1479b.dfy (hunk 2 above)
Comment (cpitclaudel) [reply Test/git-issues/git-issue-1479b.dfy:16]:
```suggestion
    launder(rep[r].(v := v)) // Error: `v` does not satisfy subtype constraints
```

File: Test/git-issues/git-issue-1479a.dfy (hunk 1 above)
Comment (cpitclaudel) [reply Test/git-issues/git-issue-1479a.dfy:13]:
```suggestion
  var dp': SubsetType := [dp][0].(t := t); // Error: `t` does not satisfy subtype constraints
```

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**